### PR TITLE
navidrome: init at 0.23.1

### PR DESCRIPTION
**Pull request by aciceri**
###### Motivation for this change
Because it's a nice application

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

This is my first pull request, hope that I've done everything correctly. Tell me if there are problems.

### CODE REVIEW

**Comment by andrea96** [comment]
This should close nixos/nixpkgs#88903

File: pkgs/servers/misc/navidrome/default.nix
Comment (prusnak) [inline pkgs/servers/misc/navidrome/default.nix:22]:
```suggestion
    platforms = [ "x86_64-linux" ];
```

File: pkgs/servers/misc/navidrome/default.nix
Comment (prusnak) [inline pkgs/servers/misc/navidrome/default.nix:21]:
```suggestion
    license = licenses.gpl3;
```

File: pkgs/servers/misc/navidrome/default.nix
Comment (prusnak) [inline pkgs/servers/misc/navidrome/default.nix:7]:
```suggestion
  src = fetchzip {
```

File: pkgs/servers/misc/navidrome/default.nix
Comment (prusnak) [inline pkgs/servers/misc/navidrome/default.nix:13]: If you use `src = fetchzip {` above, you can drop this line
```suggestion
```

File: pkgs/servers/misc/navidrome/default.nix
Comment (prusnak) [inline pkgs/servers/misc/navidrome/default.nix:1]:
```suggestion
{ stdenv, fetchzip }:
```

File: maintainers/maintainer-list.nix
Comment (prusnak) [inline maintainers/maintainer-list.nix:8814]: Keep the list of maintainers sorted

**Review by prusnak — CHANGES_REQUESTED** [review]

**Review by aciceri — COMMENTED** [review]
Thank you for your changes, I understood all of them.
However I can't "approve" them in github, even if I created the pull request, I'm not experencied with these things.
Have I to manually add them to my fork (creating a commit and then clicking on "resolve conservation" on your messages) or the changes will be automatically merged when the PR will be approved?

Moreover, a new version of navidrome is out, I also have to update the link and the hash.

**Comment by prusnak** [comment]
> However I can't "approve" them in github

You can apply the changes manually in your fork.

File: pkgs/servers/misc/navidrome/default.nix
Comment (andrea96) [reply pkgs/servers/misc/navidrome/default.nix:7]: I can't do this becuse the archive does not contain a single directory/file, if I try I get `error: zip file must contain a single file or directory`

File: pkgs/servers/misc/navidrome/default.nix
Comment (andrea96) [reply pkgs/servers/misc/navidrome/default.nix:13]: I can't do this since I can't use `fetchzip`

File: pkgs/servers/misc/navidrome/default.nix
Comment (andrea96) [reply pkgs/servers/misc/navidrome/default.nix:1]: I can't do this since I can't use `fetchzip`

File: maintainers/maintainer-list.nix
Comment (andrea96) [reply maintainers/maintainer-list.nix:8814]: I added myself at the end because the list isn't sorted, now I put my entry at the beginning in a point where the list is locally sorted. Is this what you meant?

**Comment by andrea96** [comment]
> You can apply the changes manually in your fork.
I pushed the commit with the new version and your suggested changes to my fork but at the beginning of this page, in the tab "comits" I can see only the initial commit.
Sorry for the stupid questions but this is my fist time doing these things, and thank you for reviewing this PR.

**Review by prusnak — APPROVED** [review]

File: pkgs/servers/misc/navidrome/default.nix
Comment (prusnak) [reply pkgs/servers/misc/navidrome/default.nix:21]: This has not been fixed

File: pkgs/servers/misc/navidrome/default.nix
Comment (prusnak) [reply pkgs/servers/misc/navidrome/default.nix:22]: This has not been fixed

**Review by prusnak — CHANGES_REQUESTED** [review]

**Comment by aciceri** [comment]
> This has not been fixed

@prusnak I updated to 0.23.0 but I can't understand the last two changes you suggested four days ago, aren't they already resolved?
I mean [this](https://github.com/NixOS/nixpkgs/pull/88928#discussion_r439741173) and [this](https://github.com/NixOS/nixpkgs/pull/88928#discussion_r439741190).

**Comment by prusnak** [comment]
> aren't they already resolved?
No they are not. See https://github.com/NixOS/nixpkgs/pull/88928/files

**Comment by aciceri** [comment]
> No they are not. See https://github.com/NixOS/nixpkgs/pull/88928/files

Now I marked as resolved all the changes in thae page you link. However I'm not sure to understand what you mean.
You can see [here](https://github.com/aciceri/nixpkgs/blob/navidrome/pkgs/servers/misc/navidrome/default.nix) that the description, homepage, license and platforms fields are correct (as your suggested changes of 4 days ago).

**Comment by prusnak** [comment]
>  fields are correct

No, they are not.

From your file:

```
license = stdenv.lib.licenses.gpl3;
platforms = platforms.x86_64;
```

My suggestions (you marked as resolved but you did not apply them):

```
license = licenses.gpl3;
platforms = [ "x86_64-linux" ];
```

**Comment by aciceri** [comment]
@prusnak Ahhh, I feel so stupid, forgive me, this is my first pull request on Github and I don't know all of these mechanisms. Thank you for the patience.

image

I thought that the green text meant it was fine

Now can my request be merged?

**Review by prusnak — APPROVED** [review]

**Review by rnhmjoj — CHANGES_REQUESTED** [review]
First of all, thank you for contributing to Nixpkgs!
I have a couple things to ask before merging this, sorry if this feels pedantic.
- You should separate the changes where you add yourself to the maintainers list from the "init" commit.
- Since it's not essential to keep the history of you changes in `master`, you should "squash" the later commits (0.20 -> 0.21, etc.) into the first one to make a single "init at 0.23.0" commit.

Usually this kind of things are done with an interactive rebase: `git rebase -i`, however If you are unfamiliar with git, the simplest thing to do here is:

- do a soft reset to the undo all your commits (`git rebase HEAD~6`),
- redo `git add ...` and `git commit`,
- and finally `git push --force origin` to update the remote branch

Regarding https://github.com/NixOS/nixpkgs/issues/88903#issuecomment-633736220: packages built from source are preferred for various reasons but it's ok to just fetch a binary. You are also quite fortunate because Navidrome is a go module, which is not difficult to package. Maybe for the next time you update the package you could try to built it from source. There are instructions in the manual [here](https://nixos.org/nixpkgs/manual/#ssec-go-modules).

If you have trouble with anything, just ask.

PS: nice to seen a fellow autistico, ciao!

**Comment by aciceri** [comment]
@rnhmjoj Thank to you to spend some time helping me.
I've had some problems using the commands you posted and in the end I used `git reset`, you can see in [my branch](https://github.com/aciceri/nixpkgs/commits/navidrome) that now there are only two commits. I hope that the commit messages are ok.

I can't understand why at the top of this github page there is written "6 commits", it looks like it is counting other 4 commits previous to my modifications.

> You are also quite fortunate because Navidrome is a go module, which is not difficult to package.

Initially I tried using the instructions in the manual but I failed because, if I remember correctly, 
navidrome has dependencies external to go, for example some npm packages.

> nice to seen a fellow autistico, ciao!

What a pleasant coincidence, ciao compagno!

**Comment by deluan** [comment]
Hey @aciceri, thanks for creating this! It is great to see contributions like this from the community!

I'm just chiming in to point you to the AUR package that also builds Navidrome from source:
https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=navidrome-git

Not sure how Nix packaging works but maybe you can do the same as in the AUR packaging: add a dependency to node/npm and call:
```
make setup
make buildall
```

If you need more help in building it from source, join the project's Discord and I can try to help you with this.

Another observation is that I don't see a dependency on `ffmpeg`. It is a runtime dependency required by Navidrome to work properly.

It would be great if we could automate publishing updates to this package as part of Navidrome's release pipeline. Is there any other project doing this?

**Comment by rnhmjoj** [comment]
@acieri Oops, in the first point I wrote `rebase` but I meant `reset`, sorry. Try this:

- `git rebase -i HEAD~6` (this time for real), a text editor will open
- at the top of the file, delete the lines of the other 4 commits, save and exit.
- `git push --force origin`

@deluan Hi, I don't think it will work just like that: npm will probably try to access the internet and Nix forbids that (it would not be reproducible). If I'm right a solution would be to add `node_modules` to the release tarball. This is what gitea does, which was in a similar [situation](https://github.com/go-gitea/gitea/issues/10253).

**Comment by deluan** [comment]
@rnhmjoj Releasing the `node_modules` used to build Navidrome's UI would be straightforward, but I don't see the advantage of downloading a huge tarball (445M uncompressed) to build a JS bundle of only 6.5M (uncompressed). Why not just download the JS bundle instead? What is the advantage of building this bundle during install time

Adding to that, building the GoLang server would also require downloading all libraries (go modules) used by the project from the Internet. I fail to see how this is different from what npm does...

Sorry if these may sound naive and even dumb questions, as this is my first contact with NixOS, and I don't know anything about its philosophy and how things work around here :)

And this is probably not the best place to discuss these things... :P

**Comment by rnhmjoj** [comment]
> Why not just download the JS bundle instead?

Sure, that would also work and save bandwidth too. If you can, this looks like a better idea.

> I fail to see how this is different from what npm does...

The only difference is that there is infrastructure in Nixpkgs that can handle this automatically: basically Nix gathers all the go modules required, computes a single hash and store them in a content-addressed location that can be passed to the package builder.

There is something similar for node also: it's more involved and produces tons of generated Nix expressions, so it's not great but works. I wouldn't known how to mix the two in a single build system, though.

**Comment by aciceri** [comment]
@rnhmjoj ok, I've made a terrible mess but now the git tree should be clean.

Anyway I've added the `ffmpeg` dependency as done in the [youtube-dl derivation](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/misc/youtube-dl/default.nix), it's the first package using `ffmpeg` that came to my mind.

However I remember that wen I tried to write the derivation using `buildGoModule` (or `buildGoPackage`?) I had the problem that npm needed the network connectivity, as you are saying.

i think that now the PR should be ready to be merged.

**Review by rnhmjoj — CHANGES_REQUESTED** [review]
The history looks good now, thanks.
The wrapping is not quiet right, though: `makeWrapperArgs` is an argument specific of python builder so it's actualy doing nothing in this case. You have to call `wrapProgram` manually, see for example this [package](https://github.com/NixOS/nixpkgs/blob/84cf00f98031e93f389f1eb93c4a7374a33cc0a9/pkgs/applications/window-managers/i3/blocks-gaps.nix).
Tip: to check if the program is wrapped open it with a text editor (like `vi result/bin/something`). If it's a shell script it's wrapped, otherwise it's not.

**Comment by aciceri** [comment]
I accidentally closed the PR when I deleted all the commits in my fork, now I reopened it.
I did as you told me and I divided the procedure into different phases, and checked that the output file is a shell script.

**Review by rnhmjoj — APPROVED** [review]

**Comment by rnhmjoj** [comment]
Thank you!

**Pull request merged**
